File: capacitor-plugin/android/src/main/java/it/hubagency/spatialscan/RoomRecord.kt
package it.hubagency.spatialscan

import org.json.JSONObject

/**
 * Record immutabile di una stanza scansionata.
 * Serializzato/deserializzato da RoomHistoryManager come JSON.
 */
data class RoomRecord(
    val id:            String,   // UUID generato al salvataggio
    val name:          String,   // nome assegnato dall'utente
    val timestamp:     Long,     // ms epoch (System.currentTimeMillis())
    val area:          Double,   // m²
    val height:        Double,   // m
    val wallCount:     Int,
    val openingCount:  Int,
    val floorPlanPath: String?,  // path assoluto PNG (in cacheDir)
    val glbPath:       String?   // path assoluto GLB (in cacheDir)
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("id",            id)
        put("name",          name)
        put("timestamp",     timestamp)
        put("area",          area)
        put("height",        height)
        put("wallCount",     wallCount)
        put("openingCount",  openingCount)
        if (floorPlanPath != null) put("floorPlanPath", floorPlanPath)
        if (glbPath       != null) put("glbPath",       glbPath)
    }

    companion object {
        fun fromJson(obj: JSONObject): RoomRecord = RoomRecord(
            id            = obj.getString("id"),
            name          = obj.getString("name"),
            timestamp     = obj.getLong("timestamp"),
            area          = obj.optDouble("area", 0.0),
            height        = obj.optDouble("height", 0.0),
            wallCount     = obj.optInt("wallCount", 0),
            openingCount  = obj.optInt("openingCount", 0),
            floorPlanPath = obj.optString("floorPlanPath").takeIf { it.isNotEmpty() },
            glbPath       = obj.optString("glbPath").takeIf { it.isNotEmpty() }
        )
    }
}

File: ios/Plugin/Models/RoomRecord.swift
import Foundation

struct RoomRecord: Codable {
    let id:            String
    let name:          String
    let timestamp:     Int64
    let area:          Double
    let height:        Double
    let wallCount:     Int
    let openingCount:  Int
    let floorPlanPath: String?
    let glbPath:       String?
    let openings:      [OpeningMetadata]

    init(id: String, name: String, timestamp: Int64, area: Double, height: Double,
         wallCount: Int, openingCount: Int, floorPlanPath: String?, glbPath: String?,
         openings: [OpeningMetadata] = []) {
        self.id = id; self.name = name; self.timestamp = timestamp
        self.area = area; self.height = height
        self.wallCount = wallCount; self.openingCount = openingCount
        self.floorPlanPath = floorPlanPath; self.glbPath = glbPath
        self.openings = openings
    }
}
